Ano-ify

RealOrangeOne opened the pull request:
Allow building firmware using [`ano`](https://github.com/sourcebots/Arturo), so we don't need the arduino IDE anymore!

RealOrangeOne committed:
Move submodule 1 level further

RealOrangeOne committed:
Rename firmware to cpp

RealOrangeOne committed:
Ignore ano files

RealOrangeOne committed:
Add circleci file

prophile commented on circle.yml:
```diff
@@ -0,0 +1,8 @@
+dependencies:
+    pre:
+        - sudo apt-get install arduino arduino-core arduino-mk
+        - git clone https://github.com/scottdarch/Arturo && cd Arturo && sudo make install
```
Should we move this to a submodule? Alternatively, is it available in any repos to avoid the build step?

prophile commented on .gitmodules:
```diff
@@ -1,3 +1,3 @@
 [submodule "libraries/Adafruit-PWM-Servo-Driver-Library"]
-	path = libraries/Adafruit-PWM-Servo-Driver-Library
+	path = libraries/libraries/Adafruit-PWM-Servo-Driver-Library/Adafruit-PWM-Servo-Driver-Library
```
Does this not prevent the use of the IDE?

Also this seems kind of a weird path with each component duplicated—how come this is necessary?

RealOrangeOne replied:
I believe the IDE searches nested directories.

The double `libraries` is necessary because that's the path the tool looks in, for some reason. I fear the duplicate `Adafruit...` directory was me messing up a `git mv`.

prophile replied:
Could we fix that (and see if we can drop the double-libraries?)

RealOrangeOne committed:
Revert "Move submodule 1 level further"

This reverts commit 6fbb18f340e35b04132766497a79ce4a52e5e3d7.

RealOrangeOne committed:
Remove duplicate nested path

RealOrangeOne committed:
Move servo driver to match IDE

RealOrangeOne committed:
Try fork to fix library resolution

RealOrangeOne committed:
Submodule sync

RealOrangeOne committed:
Submodule without the s

RealOrangeOne committed:
Set cpp version

RealOrangeOne committed:
Remove erroring library

This library wont build under circle, and seems to be absent locally. Rather than work out the compiler issues, we can just
delete the library. The dependency is provided by another library.

RealOrangeOne committed:
Allow downloading of firmware files, for whatever reason

RealOrangeOne committed:
fork as sourcebots

prophile commented:
Nice! Particularly enjoying the generating the firmware as build artefacts.

prophile commented on .gitmodules:
```diff
@@ -1,3 +1,3 @@
 [submodule "libraries/Adafruit-PWM-Servo-Driver-Library"]
 	path = libraries/Adafruit-PWM-Servo-Driver-Library
-	url = https://github.com/adafruit/Adafruit-PWM-Servo-Driver-Library/
+	url = http://github.com/adafruit/Adafruit-PWM-Servo-Driver-Library/
```
?

RealOrangeOne replied:
For some reason CircleCI doesn't like submodules with `https` domains, this was a nice-ish fix https://circleci.com/gh/sourcebots/servo-firmware/5

prophile commented on circle.yml:
```diff
@@ -0,0 +1,19 @@
+dependencies:
+    pre:
+        - sudo apt-get install arduino arduino-core arduino-mk
+        - sudo rm -rf /usr/share/arduino/libraries/Robot_Control
```
This is a little surprising—a comment would not go amiss :)

RealOrangeOne replied:
Theres a comment in the commit if that helps?

prophile replied:
I generally prefer those sorts of comments "this is surprising, but \<reason\>" going inline—it means that one doesn't have to trawl through `git blame` to find out why.

prophile commented on circle.yml:
```diff
@@ -0,0 +1,19 @@
+    pre:
+        - sudo apt-get install arduino arduino-core arduino-mk
+        - sudo rm -rf /usr/share/arduino/libraries/Robot_Control
+        - git clone https://github.com/sourcebots/Arturo && cd Arturo && sudo make install
```
Should we submodule this?

RealOrangeOne replied:
We could, but seeing as this is tooling rather than a build dependency, it seems better installed as one

prophile replied:
How is tooling different from a build dependency?

RealOrangeOne replied:
Tooling is needed to run it, and we'd need installed correctly locally to use. In the same way we dont submodule the arduino libraries

RealOrangeOne committed:
comment on removing library

RealOrangeOne committed:
Retry secure submodule
